### DISCUSSION TITLE
Have you considered wan2.1vace's lora model training?

### DISCUSSION THREAD

**Discussion by JioJe** [kohya-ss/musubi-tuner#313] [Ideas]
Have you considered wan2.1vace's lora model training?

**Comment by 6alvez** [comment]
Needed indeed! Anyone with any experience training for WAN2.1 VACE or FUN variants?

I am just starting with the whole training-loras deal, If I find something useful will come back with the findings.

**Reply by JioJe** [reply]
I have not found any relevant training information on the Internet, including several other training packages, such as diffusion-pipe, which do not have similar functions. Code is my weak point, looking forward to kohya

**Comment by anm-ol** [comment]
Still looking for a wan-VACE training support!
Diffusion -V does not have it and Diffsynth-Studio is very limited.

Does the Wan-Fun-control model work better than VACE?

**Reply by tavyscrolls** [reply]
IMO, Vace is the best of all 2.1 models. Zero-to-Wan is a hackable project I managed to get training on Vace, problem is I can't understand how to properly build vace_context as the paper doesn't really cover how masks are added to the 32 channels (16 each: inactive and reactive frames) and interpolated to the necessary 96 channels, so the loss spikes super high early then when it drops, the lora is basically weak noise that doesn't alter the base model much.

FWIW, if your T2V lora isn't something base Wan struggles with and you are using control types that Vace was trained to be very good at, your T2V lora will probably be ok with Vace. Some things will not work well though, 2D animation or quadrupeds for example.

**Reply by anm-ol** [reply]
The problem is that my training task depends on learning signals from the depth video, to control the generated video better.
So essentially, the training has to be V2V. 
I have skimmed through the VACE paper and the VACE training code in DiffSynth-Studio, maybe I could look at your code and drop some suggestions?

**Reply by tavyscrolls** [reply]
Just mentioned it in case you didn't know Vace is independent from the T2V side, I do understand it itself needing finetuning in cases.

It's a custom video collator in video_dataset.py I need, which needs to not only load the precached video latents, precached reference image latents, and binary masks, but use them to build vace_context. All other changes have been in train.py for VRAM optimizations, adding the vace layers as trainable layers, and pointing to the Vace classes (instead of WanModel) that were added to models/wan/modules/model.py from the same filepath found in the official Vace github, instead of the original Wan classes. The code works for training the full 1.3B model on 848x480x9 samples on 24gb (to validate, then train 14B on rented GPU) . It's a really simple training repo to modify (if poorly optimized), it's making vace_context needed here that I'm not understanding (it is 96 channels from the huggingface config.json)

```
    def forward_vace(
        self,
        x,
        vace_context,
        seq_len,
        kwargs
    ):
        # embeddings
        c = [self.vace_patch_embedding(u.unsqueeze(0)) for u in vace_context]
        c = [u.flatten(2).transpose(1, 2) for u in c]
        c = torch.cat([
            torch.cat([u, u.new_zeros(1, seq_len - u.size(1), u.size(2))],
                      dim=1) for u in c
        ])

        # arguments
        new_kwargs = dict(x=x)
        new_kwargs.update(kwargs)

        for block in self.vace_blocks:
            c = block(c, **new_kwargs)
        hints = torch.unbind(c)[:-1]
        return hints
```

**Reply by anm-ol** [reply]
I found some code, from what I understand since the Wan latents have 16 channels, out of the 96 channels, 32 come from the vace_video_latents (which is made up of reactive and non-reactive frames as explained in the technical report.), and 64 come from downsampling the mask video by 8 along H and W, hence creating 64 channels.

[This is all mentioned here in this function](https://github.com/modelscope/DiffSynth-Studio/blob/db124fa6bc63dda4401b025228c7e254b804a29f/diffsynth/pipelines/wan_video_new.py#L799)

**Reply by tavyscrolls** [reply]
Thank you much, those 2 'vace_mask_latents =' lines were exactly what I was missing and it's laid out so clearly here compared to the official repo. My code needs to be validated because I kinda hacked xformers into it, and I'm doing some tests with base Wan currently, but if this isn't added to musubi by the time I get back to Vace, I'll add lora support and fork Z-to-W so others can at least experiment.

**Reply by anm-ol** [reply]
@tavyscrolls Any plans on when you can make the code available?

**Reply by tavyscrolls** [reply]
I tried to add it but it still wasn't working, so I installed DiffSynth Studio and it doesn't even work in their repo:

`Given groups=1, weight of size [96, 3, 3, 3, 3], expected input[1, 4, 3, 242, 418] to have 3 channels, but got 4 channels instead`

I went ahead and forked the original repo if you want to take a look at it, just be aware it is still creating a shaped tensor purely for testing and you still need at least 24gb of VRAM for testing, even with LoRA. [tavyscrolls/Zero-to-WanVace](https://github.com/tavyscrolls/Zero-to-WanVace)

I shelved it a month ago to work on my current musubi project and once Wan 2.2 is supported, I will probably focus on that.
